devito/finite_differences/coefficients.py

```diff
@@ -0,0 +1,191 @@
+    function : represents the function for which the supplied coefficients
+               will be used.
+    dimension : represents the dimension with respect to which the
+                derivative is being taken.
+    coefficients : represents the set of finite difference coefficients
+                  intended to be used in place of the standard
+                  coefficients (obtained from a Taylor expansion).
+    """
+
+    def __init__(self, deriv_order, function, dimension, coefficients):
+
+        self.check_input(deriv_order, function, dimension, coefficients)
+
+        # Ensure the given set of coefficients is the correct length
+        if dimension.is_Time:
+            if len(coefficients)-1 != function.time_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+        elif dimension.is_Space:
+            if len(coefficients)-1 != function.space_order:
+                raise ValueError("Number FD weights provided does not "
+                                 "match the functions space_order")
+
+        self.is_Coefficient = True
+
+        self.deriv_order = deriv_order
+        self.function = function
+        self.dimension = dimension
+        self.coefficients = coefficients
+
+    def check_input(self, deriv_order, function, dimension, coefficients):
+        if not isinstance(deriv_order, int):
+            raise TypeError("Derivative order must be an integer")
+        if not function.is_Function:
+            raise TypeError("Coefficients must be attached to a valid function")
+        if not dimension.is_Dimension:
+            raise TypeError("Coefficients must be attached to a valid dimension")
+        # Currently only numpy arrays are accepted here.
+        # Functionality will be expanded in the near future.
+        if not isinstance(coefficients, np.ndarray):
+            raise NotImplementedError
+        return
+
+
+def default_rules(obj, functions):
+
+    def generate_subs(d):
+
+        deriv_order = d[0]
+        function = d[1]
+        dim = d[2]
+
+        if dim.is_Time:
+            fd_order = function.time_order
+        elif dim.is_Space:
+            fd_order = function.space_order
+        else:
+            # Shouldn't arrive here
+            raise TypeError("Dimension type not recognised")
+
+        side = function.get_side(dim, deriv_order)
+        stagger = function.get_stagger(dim, deriv_order)
+
+        subs = {}
+
+        indices, x0 = generate_indices(function, dim, dim.spacing, fd_order,
+                                       side=side, stagger=stagger)
+
+        coeffs = sympy.finite_diff_weights(deriv_order, indices, x0)[-1][-1]
+
+        for j in range(len(coeffs)):
+            subs.update({function.fd_coeff_symbol()
+                         (indices[j], deriv_order, function, dim): coeffs[j]})
+
+        return subs
+
+    rules = {}
```

Comment:
this is declarated a bit too prematurely , as it's populated further down in the function body

Reply:
Indeed. There was previously a conditional return statement just below this declaration but when that was no longer needed and removed I forgot to move this declaration.